crates/containerd-shim-wasm/src/sandbox/youki_instance.rs

```diff
@@ -0,0 +1,188 @@
+//! Abstractions for running/managing a wasm/wasi instance that uses youki's libcontainer library.
+
+use anyhow::Context;
+use chrono::Utc;
+use libc::{SIGINT, SIGKILL};
+use libcontainer::{
+    container::{Container, ContainerStatus},
+    signal::Signal,
+};
+use log::error;
+use nix::{
+    errno::Errno,
+    sys::wait::{waitid, Id as WaitID, WaitPidFlag, WaitStatus},
+};
+
+use crate::sandbox::{
+    instance::ExitCode,
+    instance_utils::{get_instance_root, instance_exists},
+};
+
+use crate::sandbox::InstanceConfig;
+use std::{path::PathBuf, thread};
+
+use super::{error::Error, instance::Wait, Instance};
+
+/// YoukiInstance is a trait that gets implemented by a WASI runtime that
+/// uses youki's libcontainer library as the container runtime.
+/// It provides default implementations for some of the Instance trait methods.
+/// The implementor of this trait is expected to implement the
+/// * `new_youki()`
+/// * `get_exit_code()`
+/// * `get_id()`
+/// * `get_root_dir()`
+/// * `build_container()`
+/// methods.
+#[cfg(feature = "libcontainer")]
+pub trait YoukiInstance {
+    /// The WASI engine type
+    type E: Send + Sync + Clone;
+
+    /// Create a new instance
+    fn new_youki(id: String, cfg: Option<&InstanceConfig<Self::E>>) -> Self;
```

Comment:
**suggestion / preference:** rename this to plain `new`:
* In L.65 do
```rust
<Self as YoukiInstance>::new(...)
```
* In the implementations, instead of doing
```rust
use containerd_shim_wasm::sandbox:YoukiInstance;
impl YoukiInstance for Wasi { ... }
```
do
```rust
use containerd_shim_wasm::sandbox;
impl sandbox::YoukiInstance for Wasi { ... }
```

Reply:
Done.

Reply:
That's why you would need
```rust
<Self as LibcontainerInstance>::new(...)
```

But I'm not sure that's better than the method being called `new_libcontainer`.
I guess it's just odd that the method `new` is part of a trait, and not part of the struct's impl.

I'm happy to keep this as `new_libcontainer`.